pool/pool.go

```diff
@@ -585,7 +576,130 @@ func (p *pool) removeSessionTokenAfterThreshold(cfg *callConfig) error {
 	return nil
 }
 
-func (p *pool) PutObject(ctx context.Context, params *client.PutObjectParams, opts ...CallOption) (*oid.ID, error) {
+type callContext struct {
+	// base context for RPC
+	context.Context
+
+	client Client
+
+	// client endpoint
+	endpoint string
+
+	// request signer
+	key *ecdsa.PrivateKey
+
+	// flag to open default session if session token is missing
+	sessionDefault bool
+	sessionTarget  func(session.Token)
+}
+
+func (p *pool) initCallContext(ctx *callContext, cfg *callConfig) error {
+	cp, err := p.connection()
+	if err != nil {
+		return err
+	}
+
+	ctx.key = cfg.key
+	if ctx.key == nil {
+		// use pool key if caller didn't specify its own
+		ctx.key = p.key
+	}
+
+	ctx.endpoint = cp.address
+	ctx.client = cp.client
+
+	if ctx.sessionTarget == nil && cfg.stoken != nil {
+		ctx.sessionTarget(*cfg.stoken)
+	}
+
+	// note that we don't override session provided by the caller
+	ctx.sessionDefault = cfg.stoken == nil && cfg.useDefaultSession
+
+	return err
+}
+
+type callContextWithRetry struct {
+	callContext
+
+	noRetry bool
+}
+
+func (p *pool) initCallContextWithRetry(ctx *callContextWithRetry, cfg *callConfig) error {
+	err := p.initCallContext(&ctx.callContext, cfg)
+	if err != nil {
+		return err
+	}
+
+	// don't retry if session was specified by the caller
+	ctx.noRetry = cfg.stoken != nil
+
+	return nil
+}
+
+// opens new session or uses cached one.
+// Must be called only on initialized callContext with set sessionTarget.
+func (p *pool) openDefaultSession(ctx *callContext) error {
+	cacheKey := formCacheKey(ctx.endpoint, ctx.key)
+
+	tok := p.cache.Get(cacheKey)
+	if tok != nil {
+		// use cached token
+		ctx.sessionTarget(*tok)
+		return nil
+	}
+
+	// open new session
+	var cliPrm client.CreateSessionPrm
+
+	cliPrm.SetExp(math.MaxUint32)
```

Comment:
I suppose we shouldn't use `math.MaxUint32` as expiration epoch. We have `p.stokenDuration` that should be added to the current epoch